File: FSharp/6_Coproduct.fs
module Coproduct

open FsCheck.Xunit

type Either<'TLeft, 'TRight> =
    | Left of 'TLeft
    | Right of 'TRight
    
let ind = Left
let inr = Right

type C = A of A | B of B
and A = string
and B = int

let f = C.A
let g = C.B

let ``[f; g]`` either : C =
    match either with
    | Left x -> f x
    | Right x -> g x
    
[<Property>]
let ``f is [f; g] after ind`` (x : A) =
    f x = (ind >> ``[f; g]``)  x

[<Property>]
let ``g is [f; g] after inr`` (x : B) =
    g x = (inr >> ``[f; g]``) x


let convert (coproduct, c) = 
    match coproduct with
    | Left a -> Left (a, c)
    | Right b -> Right (b, c)

[<Property>]
let left (a : int) (c : bool) =
    convert (Left a, c) = Left (a, c)

[<Property>]
let right (b : string) (c : bool) =
    convert (Right b, c) = Right (b, c)
    
// Example: convert a product (Amount, CheckNumber|CardNumber) into (Amount, CheckNumber)|(Amount, CardNumber)

File: CSharp/6_Coproduct.cs
// ReSharper disable All

using FsCheck;
using Xunit;

namespace CSharp
{
    public class Coproduct
    {
        static Either<TL, TR> ind<TL, TR>(TL l) => Either<TL, TR>.Left(l);
        static Either<TL, TR> inr<TL, TR>(TR r) => Either<TL, TR>.Right(r);

        abstract record C { }
        sealed record A(string Value) : C;
        sealed record B(int Value) : C;

        static C f(string x) => new A(x);
        static C g(int x) => new B(x);

        static C f_or_g(Either<string, int> either) => either.Match(f, g);

        [Fact]
        public void f_is_h_after_ind()
        {
            Prop.ForAll(Arb.From<string>(),
                x => f(x) == f_or_g(ind<string, int>(x)))
                .QuickCheckThrowOnFailure();
        }
        
        [Fact]
        public void g_is_h_after_inr()
        {
            Prop.ForAll(Arb.From<int>(),
                    x => g(x) == f_or_g(inr<string, int>(x)))
                .QuickCheckThrowOnFailure();
        }
    }
}